### PR TITLE
Fetch schemas index in main thread

### PR DESCRIPTION
**Pull request by saulius**
Previous attempts to collect exceptions from `Future`s proved to be difficult to manage and keep Sweeper parallel. Instead, lets simplify and fetch schema index once in the main thread before workers start.

In case of failure fails with:

```
16/06/27 11:19:39 INFO sweeper.CamusSweeper: Starting kafka sweeper
16/06/27 11:19:39 INFO sweeper.CamusSweeper: Checking schema-registry connectivity
Exception in thread "main" org.apache.http.conn.HttpHostConnectException: Connection to http://localhost:9292 refused
        at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:190)
        at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:294)
        at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:643)
        at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:479)
        at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:906)
        at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:805)
        at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:784)
        at com.linkedin.camus.sweeper.CamusSweeper.getResponse(CamusSweeper.java:663)
        at com.linkedin.camus.sweeper.CamusSweeper.testSchemaRegistryConnectivity(CamusSweeper.java:653)
        at com.linkedin.camus.sweeper.CamusSweeper.run(CamusSweeper.java:171)
        at com.linkedin.camus.sweeper.CamusSweeper.run(CamusSweeper.java:643)
        at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:70)
        at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:84)
        at com.linkedin.camus.sweeper.CamusSweeper.main(CamusSweeper.java:669)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
Caused by: java.net.ConnectException: Connection refused
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:127)
        at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180)
        ... 19 more
```

@ljank @vidma @astrauka

### CODE REVIEW

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java [hunk 1]
```diff
@@ -83,6 +89,7 @@ public CamusSweeper(Properties props)
 
   private void init()
   {
+
```
Comment (astrauka) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: ?

**Comment by astrauka** [comment]
:+1:

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java (hunk 1 above)
Comment (saulius) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: Accidental, while trying to find a good place for connectivity check.

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java [hunk 2]
```diff
@@ -222,16 +232,11 @@ public void run() throws Exception
       {
         System.err.println("unable to process " + topicName + " skipping...");
         e.printStackTrace();
-        executorService.shutdown();
-        throw e;
```
Comment (vidma) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: what kind of exceptions we might get here which do not need to be propagated?

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java [hunk 3]
```diff
@@ -280,20 +285,10 @@ private void runCollectorForTopicDir(FileSystem fs, String topic, Path topicSour
...
-    for (Future<?> task : tasksToComplete) {
-      try {
-        task.get();
-      } catch (InterruptedException e) {
-        throw e;
-      } catch (ExecutionException e) {
-        throw e;
-      }
-    }
-
```
Comment (vidma) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: don't we need to wait for the futures to complete anywhere?

> [executorService.shutdown] does not wait for previously submitted tasks to complete execution. Use awaitTermination to do that.

and we wait only for 30secs (see other comment).

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java [hunk 4]
```diff
@@ -222,16 +232,11 @@ public void run() throws Exception
...
-    if (!executorService.isTerminated()) {
-      log.info("Shutting down priority executor");
-      executorService.shutdown();
-    }
-
+    log.info("Shutting down priority executor");
+    executorService.shutdown();
     while (!executorService.isTerminated())
     {
       executorService.awaitTermination(30, TimeUnit.SECONDS);
```
Comment (vidma) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: Isn't 30 seconds  too short to wait for all topics and all paths? (they all happen async as there's no blocking inside `runCollectorForTopicDir` anymore)

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java [hunk 5]
```diff
@@ -280,20 +285,10 @@ private void runCollectorForTopicDir(FileSystem fs, String topic, Path topicSour
 
     List<Properties> jobPropsList = planner.createSweeperJobProps(topic, topicSourceDir, topicDestDir, fs);
 
-    for (Properties jobProps : jobPropsList) {
+    for (Properties jobProps : jobPropsList){
       tasksToComplete.add(runCollector(jobProps, topic));
```
Comment (vidma) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: we do not use `tasksToComplete` ArrayList anymore. so should we remove the `.add` call and the array itself?

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java (hunk 2 above)
Comment (saulius) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: I'd guess all kinds of exception from DivisionByZero to out of memory exception - you will have to contact Linkedin for more exact answer.

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java (hunk 3 above)
Comment (saulius) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: I didn't write this code, but reading documentation of `ExecutorService` really helps to understand. So the flow is as follows: `shutdown`, `while(!isNotTerminated)`, `awaitTermination`. Documentation of `isNotTerminated` is as follows:

> Returns true if all tasks have completed following shut down.

So we will be waiting for all of them to finish.

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java (hunk 5 above)
Comment (saulius) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: Lets leave as this - I just reverted to original version. Refactoring `camus` (which is a discontinued project) would be a time waste.

File: camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java (hunk 3 above)
Comment (vidma) [inline camus-sweeper/src/main/java/com/linkedin/camus/sweeper/CamusSweeper.java]: sorry, this time my fault, didn't notice the `while` loop. this should be fine.

**Comment by vidma** [comment]
👍

**Pull request merged**